service/history/historyEngine.go

```diff
@@ -2201,6 +2201,45 @@ func (e *historyEngineImpl) TerminateWorkflowExecution(
 				request.GetIdentity(),
 			)
 		})
+
+	// Terminate triggered due to commonpb.ParentClosePolicy_Terminate also includes ParentExecution info with the
+	// request. If Workflow already completed try to load current active workflow execution to allow terminate when
+	// child does continue as new.  Must compare parent execution to make sure terminate is triggered by the same parent.
```

Comment:
Every workflow contains firstExecutionRunId. I would just add this to the Terminate call and omit runID in this case. Then terminate call will always lookup a currently running workflow by WorkflowID and fail termination if firstExecutionRunId doesn't match.

Reply:
One benefit is that would work even in cases when there is no parent and can be part of the public terminate API.